sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java

```diff
@@ -0,0 +1,748 @@
+          channel.exchangeDeclare(source.spec.connectionConfig().exchange(),
+              source.spec.connectionConfig().exchangeType());
+          queueName = channel.queueDeclare().getQueue();
+          channel.queueBind(queueName, source.spec.connectionConfig().exchange(), "");
+        }
+        checkpointMark.channel = channel;
+        consumer = new QueueingConsumer(channel);
+        channel.txSelect();
+        channel.basicConsume(queueName, false, consumer);
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+      return advance();
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      try {
+        QueueingConsumer.Delivery delivery = consumer.nextDelivery();
+        if (source.spec.useCorrelationId()) {
+          String correlationId = delivery.getProperties().getCorrelationId();
+          if (correlationId == null) {
+            throw new IOException("RabbitMqIO.Read uses message correlation ID, but received "
+                + "message has a null correlation ID");
+          }
+          currentRecordId = correlationId.getBytes();
+        }
+        long deliveryTag = delivery.getEnvelope().getDeliveryTag();
+        checkpointMark.sessionIds.add(deliveryTag);
+        current = delivery.getBody();
+        currentTimestamp = new Instant(delivery.getProperties().getTimestamp());
+        if (currentTimestamp.isBefore(checkpointMark.oldestTimestamp)) {
+          checkpointMark.oldestTimestamp = currentTimestamp;
+        }
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+      return true;
+    }
+
+    @Override
+    public void close() throws IOException {
+      if (channel != null) {
+        try {
+          channel.close();
+        } catch (Exception e) {
+          // ignore
+        }
+      }
+      if (connection != null) {
+        connection.close();
+      }
+    }
+
+  }
+
+  /**
+   * A {@link PTransform} to publish messages to a RabbitMQ server.
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<byte[]>, PDone> {
+
+    @Nullable abstract ConnectionConfig connectionConfig();
+
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setConnectionConfig(ConnectionConfig connectionConfig);
+      abstract Write build();
+    }
+
+    /**
+     * Define the {@link ConnectionConfig} to RabbitMQ server.
+     */
+    public Write withConnectionConfig(ConnectionConfig connectionConfig) {
+      checkArgument(connectionConfig != null, "connectionConfig can not be null");
+      return builder().setConnectionConfig(connectionConfig).build();
+    }
+
+    public Write withUri(String uri) {
+      checkArgument(uri != null, "uri can not be null");
+      return builder().setConnectionConfig(connectionConfig().withUri(uri)).build();
+    }
+
+    public Write withQueue(String queue) {
```

Comment:
In what cases does a producer need to specify a single queue?

Reply:
In the queuing behavior use case (with concrete queue) instead of exchange use case.